Src/GoogleApis.Core/Apis/Util/PageStreamer.cs

```diff
@@ -0,0 +1,137 @@
+    /// <typeparam name="TResponse">The type of response obtained when fetching pages</typeparam>
+    /// <typeparam name="TToken">The type of the "next page token", which must be equatable to itself</typeparam>
+    public sealed class PageStreamer<TResource, TRequest, TResponse, TToken>
+        where TToken : IEquatable<TToken>
+    {
+        private readonly Func<TRequest, TToken, TRequest> _requestProvider;
+        private readonly Func<TResponse, TToken> _tokenExtractor;
+        private readonly Func<TResponse, IEnumerable<TResource>> _resourceExtractor;
+        private readonly TToken _emptyToken;
+
+        /// <summary>
+        /// Creates a paginator for later use.
+        /// </summary>
+        /// <param name="requestProvider">Function to combine an initial request and a page token
+        /// into a new request. Must not be null.</param>
+        /// <param name="tokenExtractor">Function to extract the next page token from a response.
+        /// Must not be null.</param>
+        /// <param name="resourceExtractor">Function to extract a sequence of resources from a response.
+        /// Must not be null.</param>
+        /// <param name="emptyToken">The token to use to detect when there are no more pages available.</param>
+        public PageStreamer(
+            Func<TRequest, TToken, TRequest> requestProvider,
+            Func<TResponse, TToken> tokenExtractor,
+            Func<TResponse, IEnumerable<TResource>> resourceExtractor,
+            TToken emptyToken)
+        {
+            if (requestProvider == null)
+            {
+                throw new ArgumentNullException("requestProvider");
+            }
+            if (tokenExtractor == null)
+            {
+                throw new ArgumentNullException("tokenExtractor");
+            }
+            if (resourceExtractor == null)
+            {
+                throw new ArgumentNullException("resourceExtractor");
+            }
+            this._requestProvider = requestProvider;
+            this._tokenExtractor = tokenExtractor;
+            this._resourceExtractor = resourceExtractor;
+            this._emptyToken = emptyToken;
+        }
+
+        /// <summary>
+        /// Lazily fetches resources a page at a time.
+        /// </summary>
+        /// <param name="initialRequest">The initial request to send. If this contains a page token,
+        /// that token is maintained. This is also passed to the request provider along with a token
+        /// to provide another request. (The provider may build a new object, or modify the existing request.)</param>
+        /// <param name="responseFetcher">A function to fetch a response containing a page of results given a request.</param>
+        /// <returns>A sequence of resources, which are fetched a page at a time. Must not be null.</returns>
+        public IEnumerable<TResource> Fetch(TRequest initialRequest, Func<TRequest, TResponse> responseFetcher)
```

Comment:
If you can provide examples of APIs in the framework that use it, I'll consider it more... That's definitely an interesting page, but it's not something I've seen in reality.

(It also introduces the question about what other sync operations should take a cancellation token. Any network request could be regarded as long-running...)